### DISCUSSION TITLE
Performance bottlenecks and findings

### DISCUSSION THREAD

**Discussion by Boshen** [rome/tools#2308] [General]
A summary of perf bottlenecks and findings. 
(I don't have time for any code work so I'll just jot down the things I find.) 

General background research study:
* follow the work of https://github.com/nnethercote, read his [blog](https://nnethercote.github.io/), read the [perf book](https://nnethercote.github.io/perf-book/title-page.html)

**Comment by Boshen** [comment]
Bounded by Memory Allocation:
* switch to jemallocator: https://nnethercote.github.io/perf-book/heap-allocations.html#using-an-alternative-allocator
* use stack allocated string where possible: https://github.com/epage/string-benchmarks-rs
  * https://github.com/rome/tools/blob/76bf61e5f8c4a279c1bb7fb640710b521aac1b10/crates/rome_js_parser/src/syntax/binding.rs#L121
 * use [smallvec](https://docs.rs/smallvec/latest/smallvec/index.html) where possible?
* use [memory arena](https://docs.rs/bumpalo/latest/bumpalo/collections/vec/struct.Vec.html) for Parser.events?
  * https://github.com/rome/tools/blob/main/crates/rome_js_parser/src/parser.rs#L72

**Reply by leops** [reply]
Being on Windows I couldn't try `jemalloc` but I did have a branch switching to `mimalloc` and if I remember correctly there was a ~10% improvement to the performances (but it might have included the formatter as well which puts even more stress on the allocator than the parser does)

**Reply by MichaReiser** [reply]
Thanks @Boshen for looking into all these optimisations! 

Memory arenas are certainly interesting but I'm not sure if we'll see a huge improvement for parse-events because the parse events are all stored in a single vector. However, I think it could be very interesting to use arenas for our `Green*` data structures because building these up (and de-allocating them) takes up a considerable amount of time.

**Reply by Boshen** [reply]
For future reference, this [PR](https://github.com/rome/tools/pull/2634) improved the formatter performance by ~5-20%, build time by 50%. 

> It mainly reduces the number of small transient object (~128kb) from 200k to 80k. My interpretation is that this is because it's no longer necessary to allocate a vector for every format functions call that return more than one element.

There are more performance improvements to be made:
* All the "nesting" IR elements still allocate a vector internally or Box a FormatElement. It should be possible to re-write our IR to a flat IR that uses Start and End markers and everything in between would be the content of that element. It should then become possible to only do a single allocation for formatting the whole document which should reduce another massive amount of allocations.
* Refactor member chain, conditional, and binary expression formatting to not use VecBuffer internally anymore

**Reply by Boshen** [reply]
> Being on Windows I couldn't try `jemalloc` but I did have a branch switching to `mimalloc` and if I remember correctly there was a ~10% improvement to the performances (but it might have included the formatter as well which puts even more stress on the allocator than the parser does)

~50-200% improvement done in this [PR](https://github.com/rome/tools/pull/2900)

**Reply by Conaclos** [reply]
[ArrayVec](https://crates.io/crates/arrayvec) when possible may be a better choice than SmallVec. It does not use a fallback to heap allocation.

**Reply by ematipico** [reply]
> [ArrayVec](https://crates.io/crates/arrayvec) when possible may be a better choice than SmallVec. It does not use a fallback to heap allocation.

Do you have some link that explains why this crate is a better choice than `smallvec`?

**Reply by Conaclos** [reply]
@ematipico

The [heap allocation section](https://nnethercote.github.io/perf-book/heap-allocations.html#short-vecs)  from the perf book talks about that.

**Reply by IWANABETHATGUY** [reply]
> @ematipico
> 
> The [heap allocation section](https://nnethercote.github.io/perf-book/heap-allocations.html#short-vecs) from the perf book talks about that.

Notice the prerequisite, 
> If you have many short vectors and you precisely know their maximum length, ArrayVec from the [arrayvec](https://crates.io/crates/arrayvec) crate is a better choice than SmallVec. It does not require the fallback to heap allocation, which makes it a little faster. [Example](https://github.com/rust-lang/rust/pull/74310/commits/c492ca40a288d8a85353ba112c4d38fe87ef453e).

**If you have many short vectors and you precisely know their maximum length.** The Identifier could have any length less than `std::usize::MAX` in theory.

**Comment by Boshen** [comment]
rustc options:
In Cargo.toml:

```toml
[profile.release]
lto = "fat"
codegen-units = 1
panic = "abort"
strip = "debuginfo"
```

In `.cargo/config.toml`

```toml
[build]
rustflags = ["-C", "target-cpu=native"]
```

**Reply by leops** [reply]
The `lto` and `strip` options are already set as part of the release workflow (https://github.com/rome/tools/blob/main/.github/workflows/release.yml#L108-L111), I didn't set `codegen-units` to one because I thought it was the default for the release profile but I might have been wrong. For the others, I'm not really sure about aborting on panics because while it may improve performance it could also degrade the user experience in case of a bug, with unwinding we can at least try to recover from these errors. Finally setting the `target-cpu` to native feels like cheating, it may improve performance if the benchmarks are being run on the same machine that builds them but in practice we distribute prebuilt binaries to the users, so those have to be optimized for the common case.

**Comment by Boshen** [comment]
Bounded by string comparison:
* https://github.com/rome/tools/blob/76bf61e5f8c4a279c1bb7fb640710b521aac1b10/crates/rome_js_parser/src/syntax/binding.rs#L72
  * This one is small but anything on the identifier path can impact perf
 
I can't tell if [String Interning](https://crates.io/crates/string_cache) is needed for this project.

**Reply by leops** [reply]
I did try to add interning to identifiers in the parser, but interestingly the results I got was that it made the performance of the parser generally worse so I didn't pursue this idea further. Although I still think we will have to implement some form of interning eventually because down the line linters / analyzers will needs a cheap and efficient way to lookup and compare names in a program.

**Reply by MichaReiser** [reply]
@leops that's an interesting point and again raises the discussion of symbol resolution should or should not happen inside of the parser. 

Regarding interning: Lookups in the analyzers would ideally work with a `Symbol` type (rather than operating on strings directly) because a name by itself isn't unique enough. It also requires knowledge about the scope. The `Symbol` type could then act as cheap "id" to lookup and compare symbols.

**Reply by ematipico** [reply]
Rome classic did an interesting decision. It decided to leave the symbol resolution out of the parsing. The symbol resolution was done during the the "compiling step".  There's this `CompilerContext` class which was in charge of taking a valid AST and creating the bindings out of it. Then an analyzer would get this `CompilerPath` (a wrapper around the `CompilerContext`) and it would be able to get the bindings as needed: https://github.com/rome/tools/blob/archived-js/internal/compiler/lint/rules/js/noUnusedVariables.ts#L73

**Reply by Stupremee** [reply]
I can't tell anything about the performance improvement interning would get because we never tried it in rslint. However, if you're going to implement it, I recommend using the [lasso](https://lib.rs/lasso) because it's by far the best string interning library IMO. It outperforms most other interning crates, provides really cool and optionally multithreaded APIs. Another crate you can have a look at is [cstree](https://lib.rs/cstree) which is basically rowan with improvements, and especially they have string interning built in.

**Comment by Boshen** [comment]
Lexer: anything we can improve on the lexer will have the biggest impact.
* I doubt that `unwind_loop` has actual perf can, since that's from 2 years ago, rust and llvm should optimize things and this `unwind_loop` will block it? (This macro also bloats binary output size)

**Reply by MichaReiser** [reply]
It should be straightforward to remove `unwind_loop` to see if it changes the performance in a meaningful way.

I'm not entirely convinced that the lexer is the area where we can achieve the most speedup (except maybe using SIMD?). Old benchmarks reported the numbers by time spent lexing, parsing, and in the tree sink (maybe something we could add again with tracing @leops ?). These numbers showed that lexing is often the cheapest of all and the tree-sink makes up about 50% of the time.

**Reply by MichaReiser** [reply]
Related PR #2311

**Reply by Boshen** [reply]
In this [PR](https://github.com/rome/tools/pull/2600), performance for identifier lookup is improved with a [lookup table](https://github.com/Boshen/unicode-id-start/blob/master/src/tables.rs), resulting a ~5% improvement, and build size reduced by 2kb.

**Comment by Boshen** [comment]
I see there are a lot of `assert_eq!`, is it possible to replace them with `debug_assert_eq!` so they can be stripped away for release build?

**Reply by Boshen** [reply]
https://github.com/rome/tools/pull/2379#issuecomment-1093704797
No significant changes 😢
